libnavigation-core/src/main/java/com/mapbox/navigation/core/HistoryRecordingStateHandler.kt

```diff
@@ -0,0 +1,104 @@
+package com.mapbox.navigation.core
+
+import androidx.annotation.UiThread
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.core.internal.HistoryRecordingStateChangeObserver
+import com.mapbox.navigation.core.trip.session.NavigationSessionState
+import com.mapbox.navigation.core.trip.session.NavigationSessionUtils
+import com.mapbox.navigation.core.trip.session.TripSessionState
+import com.mapbox.navigation.core.trip.session.TripSessionStateObserver
+import com.mapbox.navigation.utils.internal.logI
+
+private const val TAG = "HistoryRecordingStateHandler"
+
+@UiThread
+internal class HistoryRecordingStateHandler(
+    private var currentState: NavigationSessionState
+) : TripSessionStateObserver {
+
+    private val historyRecordingStateChangeObservers =
+        mutableSetOf<HistoryRecordingStateChangeObserver>()
```

Comment:
The class is marked as `UiThread`. And the observer invocation will be from UiThread. If I wanted to make this class thread-safe, I'd also have to synchronise access to all the properties. I decided that it's an unnecessary complication since all the interactions with this class are from the main thread.

Reply:
Agree, but the main concern for me is the case when the observer unregisters itself. I mean something like this:
```kotlin
object : HistoryRecordingStateChangeObserver {
    override fun onShouldStartRecording(state: NavigationSessionState) {
        stateHandler.unregisterStateChangeObserver(this)
    }
}
```
Even if everything happens on the main thread, you will get `ConcurrentModificationException` in this case. `CopyOnWriteArraySet` fixes this for you.

Reply:
Oh, I see. For this particular case I don't see why the observer would unregister itself, but it still makes sense. Changed to CopyOnWriteArraySet.